### PR TITLE
[CARBONDATA-431]bigdecimal compression.

### PR DESCRIPTION
**Pull request by ashokblend**
This pr will do compression for bigdecimal data type as carbon does for double datatype. It is in progress.

### CODE REVIEW

**Comment by chenliang613** [comment]
@ashokblend  please create dev mailing list discussion and raise Apache JIRA ISSUE in advance.

**Comment by ravipesala** [comment]
Build finished.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.5.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/69/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.5.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/70/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.5.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/75/

File: core/src/main/java/org/apache/carbondata/core/compression/BigDecimalCompressor.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    Long[] maxValues = (Long[]) maxValue;
+    Object leftCompressedValue = getCompressedValues(
+        bigdCompressionFinder.getLeftCompType(), dataHolder,
+        bigdCompressionFinder.getLeftChangedDataType(), maxValues[0], 0);
+    readLeft = false;
+    Object rightCompressedValue = getCompressedValues(
+        bigdCompressionFinder.getRightCompType(), dataHolder,
+        bigdCompressionFinder.getRightChangedDataType(), maxValues[1], 0);
+    return new Object[] { leftCompressedValue, rightCompressedValue };
+
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/compression/BigDecimalCompressor.java]: remove empty line

File: core/src/main/java/org/apache/carbondata/core/compression/BigIntCompressor.java [hunk 2]
```diff
@@ -44,6 +44,11 @@ protected Object compressMaxMin(DataType changedDataType, CarbonWriteDataHolder
       Object max) {
     long maxValue = (long) max;
     long[] value = dataHolder.getWritableLongValues();
+    return compressMaxMin(changedDataType, maxValue, value);
+  }
+
+  protected Object compressMaxMin(DataType changedDataType, long maxValue,
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/compression/BigIntCompressor.java]: please add description to these compress functions, make it easier to understand by reader

File: core/src/main/java/org/apache/carbondata/core/util/CompressionFinder.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+  }
+
+  /**
+   * Compression type priority.
+   * ACTUAL is the highest priority and DIFFNONDECIMAL is the lowest
+   * priority
+   */
+  enum PRIORITY {
+    /**
+     *
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/util/CompressionFinder.java]: remove unnecessary comment

File: core/src/main/java/org/apache/carbondata/core/util/ValueCompressionUtil.java [hunk 4]
```diff
@@ -138,8 +141,7 @@ public static CompressionFinder getCompressionFinder(Object maxValue, Object min
     // ''b' for decimal, 'l' for long, 'n' for double
     switch (measureStoreType) {
       case 'b':
-        return new CompressionFinder(COMPRESSION_TYPE.BIGDECIMAL, DataType.DATA_BYTE,
-            DataType.DATA_BYTE);
+        return getBigDecimalCompressorFinder(maxValue, minValue, 0, dataTypeSelected);
```
Comment (piaoyats) [inline core/src/main/java/org/apache/carbondata/core/util/ValueCompressionUtil.java]: if mantissa equals to 0 hear for big decimal? if it is, keep passing mantissa to getBigDecimalCompressorFinder

File: core/src/main/java/org/apache/carbondata/core/util/ValueCompressionUtil.java [hunk 5]
```diff
@@ -196,6 +198,31 @@ private static CompressionFinder getDoubleCompressorFinder(Object maxValue, Obje
     }
   }
 
+  private static CompressionFinder getBigDecimalCompressorFinder(
+      Object maxValue, Object minValue, int decimal, byte dataTypeSelected) {
```
Comment (piaoyats) [inline core/src/main/java/org/apache/carbondata/core/util/ValueCompressionUtil.java]: change decimal to mantissa, keep same with others

File: core/src/main/java/org/apache/carbondata/core/util/ValueCompressionUtil.java [hunk 6]
```diff
@@ -243,46 +270,64 @@ public static Object getCompressedValues(COMPRESSION_TYPE compType, double[] val
...
-      }
+  /**
+   *
+   * @param compressionFinder for measure other then bigdecimal
+   * @return
+   */
+  private static UnCompressValue getUncompressedValue(
+      CompressionFinder compressionFinder) {
+    if (compressionFinder instanceof BigDecimalCompressionFinder) {
+      return getUncompressedValue(
```
Comment (piaoyats) [inline core/src/main/java/org/apache/carbondata/core/util/ValueCompressionUtil.java]: same instanceof here

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.5.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/134/

File: core/src/main/java/org/apache/carbondata/core/util/BigDecimalCompressionFinder.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+public class BigDecimalCompressionFinder extends CompressionFinder {
+
+  private COMPRESSION_TYPE leftCompType;
+
+  private COMPRESSION_TYPE rightCompType;
+
+  private DataType leftActualDataType;
+
+  private DataType rightActualDataType;
```
Comment (ashokblend) [inline core/src/main/java/org/apache/carbondata/core/util/BigDecimalCompressionFinder.java]: ya it means actual datatype, have updated comment for more clarity

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.5.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/137/

**Comment by ashokblend** [comment]
review comments are handled, please check

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.5.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/139/

**Comment by jackylk** [comment]
LGTM
Thanks for working for this

**Pull request closed without merging**